src/Dev/Validation/RelationValidationService.php

```diff
@@ -0,0 +1,639 @@
+<?php
+
+namespace SilverStripe\Dev\Validation;
+
+use ReflectionException;
+use SilverStripe\Core\ClassInfo;
+use SilverStripe\Core\Config\Configurable;
+use SilverStripe\Core\Flushable;
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\Core\Resettable;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\DB;
+
+/**
+ * Class RelationValidationService
+ *
+ * Basic validation of relationship setup, this tool makes sure your relationships are setup correctly in both directions
+ * The validation is configurable and inspection can be narrowed down by namespace, class and relation name
+ *
+ * This tool runs automatically via flush and outputs notices
+ * For strict validation it is recommended to hook this up to your unit test suite
+ */
+class RelationValidationService implements Flushable, Resettable
+{
+
+    use Configurable;
+    use Injectable;
+
+    /**
+     * Only inspect classes with the following namespaces/class prefixes
+     * Empty string is a special value which represents classes without namespaces
+     * Set value to null to disable the rule (useful when overriding configuration)
+     *
+     * @var array
+     */
+    private static $allow_rules = [
+        'empty' => '',
+        'app' => 'App',
+    ];
+
+    /**
+     * Any classes with the following namespaces/class prefixes will not be inspected
+     * This config is intended to be used together with @see $allow_rules to narrow down the inspected classes
+     * Empty string is a special value which represents classes without namespaces
+     * Set value to null to disable the rule (useful when overriding configuration)
+     *
+     * @var array
+     */
+    private static $deny_rules = [];
+
+    /**
+     * Relations listed here will not be inspected
+     * Format is <class>.<relation>
+     * for example: Page::class.'.LinkTracking'
+     *
+     * @var array
+     */
+    private static $deny_relations = [];
+
+    /**
+     * Ignore any configuration, useful for debugging specific classes
+     *
+     * @var bool
+     */
+    protected $ignoreConfig = false;
+
+    /**
+     * @var array
+     */
+    protected $errors = [];
+
+    public function flushErrors(): void
+    {
+        $this->errors = [];
+    }
+
+    public static function reset(): void
+    {
+        $service = self::singleton();
+        $service->flushErrors();
+        $service->ignoreConfig = false;
+    }
+
+    /**
+     * @throws ReflectionException
+     */
+    public static function flush(): void
+    {
+        self::singleton()->executeValidation();
```

Comment:
Doesn't seem like this is configurable, instead it just runs on default.  I think this needs to configurable with a default of off so that it's opt-it.   Silverstripe runs just fine with missing back-relations so I think we should leave it up to developers if they want this additional linting enabled or not